src/builder.rs

```diff
@@ -177,45 +176,75 @@ impl<'ctx> Builder<'ctx> {
         let c_string = to_c_str(name);
         let mut args: Vec<LLVMValueRef> = args.iter().map(|val| val.as_value_ref()).collect();
 
-        #[cfg(not(any(
-            feature = "llvm4-0",
-            feature = "llvm5-0",
-            feature = "llvm6-0",
-            feature = "llvm7-0",
-            feature = "llvm8-0",
-            feature = "llvm9-0",
-            feature = "llvm10-0",
-            feature = "llvm11-0",
-            feature = "llvm12-0",
-            feature = "llvm13-0",
-        )))]
         let value = unsafe {
-            let fn_ty_ref = callable_value.as_type_ref();
-            LLVMBuildCall2(
+            LLVMBuildCall(
                 self.builder,
-                fn_ty_ref,
                 fn_val_ref,
                 args.as_mut_ptr(),
                 args.len() as u32,
                 c_string.as_ptr(),
             )
         };
 
-        #[cfg(any(
-            feature = "llvm4-0",
-            feature = "llvm5-0",
-            feature = "llvm6-0",
-            feature = "llvm7-0",
-            feature = "llvm8-0",
-            feature = "llvm9-0",
-            feature = "llvm10-0",
-            feature = "llvm11-0",
-            feature = "llvm12-0",
-            feature = "llvm13-0",
-        ))]
+        unsafe { CallSiteValue::new(value) }
+    }
+
+    /// Builds a function call instruction.
+    /// [`FunctionValue`]s can be implicitly converted into a [`CallableValue`].
+    /// See [`CallableValue`] for details on calling a [`PointerValue`] that points to a function.
+    ///
+    /// [`FunctionValue`]: crate::values::FunctionValue
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use inkwell::context::Context;
+    ///
+    /// // A simple function which calls itself:
+    /// let context = Context::create();
+    /// let module = context.create_module("ret");
+    /// let builder = context.create_builder();
+    /// let i32_type = context.i32_type();
+    /// let fn_type = i32_type.fn_type(&[i32_type.into()], false);
+    /// let fn_value = module.add_function("ret", fn_type, None);
+    /// let entry = context.append_basic_block(fn_value, "entry");
+    /// let i32_arg = fn_value.get_first_param().unwrap();
+    /// let md_string = context.metadata_string("a metadata");
+    ///
+    /// builder.position_at_end(entry);
+    ///
+    /// let ret_val = builder.build_call(fn_type, fn_value, &[i32_arg.into(), md_string.into()], "call")
+    ///     .try_as_basic_value()
+    ///     .left()
+    ///     .unwrap();
+    ///
+    /// builder.build_return(Some(&ret_val));
+    /// ```
+    #[llvm_versions(14.0..=latest)]
+    pub fn build_call<F>(
+        &self,
+        fn_ty: FunctionType<'ctx>,
+        function: F,
```

Comment:
another api design remark: I think this `fn_ty` is needed in the case that ` F` is a function pointer. However, usually we call functions. That means that all function calls now need to pass an extra `fun.get_type()` argument. 

I'd prefer an api where calling a function is as simple as it is today, and only when you deal with function pointers do you have to explicitly provide the type.

Reply:
I believe it's now each particular's FE's responsibility to store it's function types.
This crate uses the underlying llvm_sys API, so it's not possible without a huge redesign. And even then the resulting LLVM behavior is unclear.

Reply:
really? I'd think that a function has a known type, since it is a global value.

Reply:
On https://llvm.org/docs/OpaquePointers.html, under Migration Instructions

> For calls, use getFunctionType().

that suggests to me that using `FunctionValue::get_type` is safe and in fact the recommended way to supply a function type to `call`

Reply:
It's probably about the `call` instruction (operation) value, not it's function argument.
Even if it's redundant, how would calls to function pointers work?

Reply:
As far as I remember `get_type` stopped working literally everywhere, so I had to store types myself for all calls and invokes on my FE.
Anyway I'm glad that I have moved all the type logic to my FE so it's now completely unified and clean. So I prefer this approach.